### DISCUSSION TITLE
How do you deal with child component model interactions?

### DISCUSSION THREAD

**Discussion by ju5t** [livewire/livewire#1413] [4. General]
I have a poor Vue component that's part of our order views. I'm going to rewrite it in Livewire. 

The order needs to:

- Have a list of articles that can be added and removed.
- Some basic model interaction, e.g. the customer the order is for, state, you name it.
- The ability to save.

The Vue component only does the dynamic adding and removing of articles. This can be rewritten in Livewire with some child components. This isn't very difficult and I won't go into specifics here.

My question is about what's next. There are two situations.

- A new order.
- An existing order.

## A new order

It's easy to add a new order when a user hits save. It's like riding a bike.

## An existing order

Updating the fields of the order is straight forward too. 

What I'm not sure about are the articles. From a UX point of view you would like the entire view to behave the same. Users have to specifically save the order, it shouldn't hide this behaviour.

What do you do? Do you 'just' save the order (this is the current behaviour of our Vue component)? Do you throw away the articles and recreate them? Or is there another option?

**Comment by mblarsen** [comment]
I'm not 100% sure what you are asking. So please ignore my reply if it doesn't apply.

So what I think you are pointing out is that in Vue and client-heavy applications you create the articles in memory and only when you are satisfied with your order you "place it" which means that you persist it to the database through the API. But because Livewire in a sense "collapses the stack" there is in a sense only the backend and you are hesitant to create the order if it doesn't push through?

Two points to be made (maybe more, hehe). 

1. It is common in ecommerce to separate the actual order from the cart, what is in cart is often just called that the cart, or it can be called an offering. Using this approach you actually do just save the data and load it from the DB again on page refresh and what not. Render it with blade and so on.

2. Practically Livewire is still a front-end framework, it just helps you sync the "requests" the "API" that is your component and its public properties. So when you program your Livewire component, think of the props as Vue props and not as the final data being saved. It is up to you if you want to persist the articles. You can store them in the database or in the session. It really doesn't matter.

This is overly simplified how Livewire "replaces" vue.

|Vue|Livewire front-end|Livewire controller/api part|
|---|---|---|
|props|public properties on component| |
|computed|public getXXXProperty| |
|methods|public methods| |
|api call| |public methods but the payload is your public props on the component|
|events|$listeners|$listeners|

**Comment by ju5t** [comment]
Hi @mblarsen, thanks for the very extensive answer. Sorry if I haven't made myself clear.

First of all, frontend and backend always confuse me. To clarify, when I talk about frontend in this topic, I mean what the customer can see. When I talk about backend I mean what a sales person working for the company owning the app can see.

It's not frontend based. There is no cart. It's much more CRUD-like. A sales person will open the create view and starts filling the details. They will hit save and the order gets saved by Livewire, it's in a particular state (e.g. waiting for payment) and the articles are saved too. The sales person is redirected to page XYZ.

The customer is notified about the order and want's to make changes. The customer calls the sales person and an article has to be removed and two other articles need to be added. The sales person makes the change and hits save and now it's up to Livewire again. We're now at the point where my question is about.

How would you and other developers approach this?

1. Update the order, remove all existing articles and add a new set of articles stored in Livewire?
2. Update the order, do some comparison, remove the articles that are removed and add the new ones?
3. Or something entirely else, instantly updating and removing articles from the database and update the order when the sales person hits save?

Up's and down's:

- The upside of 1 is that it's really easy. The downside is that you lose your timestamps.
- The upside of 2 is that you keep your timestamps. The downside is that it's more complex, requiring more queries.
- The upside of 3 is that it's also fairly easy. The downside is that new and existing orders are treated differently and it may confuse the sales people.

I hope things are a little clearer now.

**Comment by mblarsen** [comment]
Thanks for the explanation it is much clearer now :D

Okay so there are two cases you have to consider:

1. an article data is changed, price, quantity or whatever, the point is it is something on that article on the order.
2. an article is added or removed.

To deal with the first point you can make use of model binding. However, currently Livewire does not support collections of objects to be used as model bindings. This may change in the future :crossed_fingers:. 

But you can turn the collection into an array and binging works. So make sure you are following this approach.

```
class OrderEditor extends Compoent
{
    public $order;
    public $articles;

    public function mount(Order $order)
    {
        $this->order = $order
        $this->articles = $order->articles->toArray();
    }
}
```

```
@foreach ($articles as $index => $article)
<input wire:model="articles.{{$i}}.qualtity" />
@endforeach
```
Maybe you have this already. Anyway what this does is binding into the article in the array. This ensures that for our existing articles we track the changes.

You can add validation in `updated()` as needed using regular Laravel validation with keys like `articles.*.quantity => min:1`.

So onto the 2nd point. You probably have some widget to find or add new articles. Once you click add you invoke a method on the component, either directly with an ID to the article or using an event that or component listens to.

So there are too approaches now. Remember just by changing the quantity of an existing article we haven't actually saved the changes. Let's first assume that we don't save either. So on to adding a new article.

```php
    public function addArticle($id)
    {
        $this->order->articles()->save(new Article(['product_id' => $id, 'quantity' => 1]);
     }
```

At this point the article is linked to the order in the database. Your model is probably different but that is irrelevant.
However it isn't included in our array. So we have to update the array. 

```php
        $this->articles = $order->articles->toArray();
```
But this overrides our unsaved quantity change. So either we have to save first. You can do that in `updated()` by running foreach/each on each of the associative arrays in $this->articles.

```php
foreach ($this->artticles as $article) {
  Article::where($article['id'])->update($article);
}
```

or we could update the array instead of replacing it.

```php
        $this->articles = collect($this->articles)->add($newArtilce->toArray())->toArray();
```

So that covers updating and adding. Lastly to remove you have to follow your decision of when to save. If you save en `updated()` you also remove there. If you have a final `save()` function then you do it there. Basically just pluck the id of the collect($this->articles)->pluck('id') and intersect it with $this->order->articles->pluck('id'). The ones that are left you delete.

---
But your question was broader. Should you update existing or replace all is what it boils down to. It really  depends on the system. Does it matter that you create a new line every time? Are there any reasons to keep existing around?

If there are more than one sales person and the prices are not fixed it may be of interest to keep the articles around as the price may be different and you are also want to [audit](http://www.laravel-auditing.com/) the articles and the order.

If none of that matters just do what is simplest.

**Comment by mblarsen** [comment]
Anyway, I don't think this is really Livewire related. At least you'd be faced with the same questions if it was Vue, React, Express, Django :)

**Reply by ju5t** [reply]
You're right, it doesn't matter which frontend you throw at it, apart from Django, as Django ≈ Laravel :-)

You gave me some great tips. For one I didn't know you could do `article.{{ id }}.quantity`. I wish we never bothered with Vue. Livewire is so much cooler if you're working with Laravel.

I'm going with updating the array and remove the items on `save()`.

Thanks for the extensive replies!

**Reply by mblarsen** [reply]
Cool, good luck. Hope it works out.
